compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAccessToAcquireRelease.cpp

```diff
@@ -4,88 +4,110 @@
-              logicalObjectFifoToDma[accessOp.getInput()].getResult(),
-              LogicalObjectFifoPort::Consume);
-          auto newAccessOp = rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
-              rewriter.getUnknownLoc(), acquireOp.getResult(),
-              AMDAIE::MemoryAccess::Read);
-          rewriter.replaceAllUsesWith(accessOp.getResult(),
-                                      newAccessOp.getResult());
-          logicalObjectFifoToLastAccess[accessOp.getInput()] = accessOp;
-          return WalkResult::advance();
-        });
-    if (res.wasInterrupted()) return failure();
+    for (auto &&[logicalObjectFifo, accessOps] : fifosToAccesses) {
+      for (uint64_t i = 0; i < accessOps.size(); ++i) {
+        AMDAIE::LogicalObjectFifoAccessOp accessOp = accessOps[i];
 
-    // Insert release for remaining read access operations at end of block.
-    for (auto &&[value, accessOp] : logicalObjectFifoToLastAccess) {
-      Block *parentBlock = accessOp->getBlock();
-      if (!parentBlock->back().hasTrait<OpTrait::IsTerminator>()) {
-        rewriter.setInsertionPointToEnd(parentBlock);
-      } else {
-        rewriter.setInsertionPoint(parentBlock->getTerminator());
-      }
-      if (!logicalObjectFifoToDma.contains(accessOp.getInput())) {
-        accessOp.emitOpError()
-            << "read access not found as source of DMA operation";
-        return failure();
+        Value input = accessOp.getInput();
+        if (!logicalObjectFifoToConnection.contains(input)) {
+          return accessOp.emitOpError()
+                 << "does not have a connection in the logicalobjectfifo map";
+        }
+
+        // Insert the access op.
+        rewriter.setInsertionPoint(accessOp);
+        Block *block = accessOp->getBlock();
+        auto acquireOp = rewriter.create<AMDAIE::LogicalObjectFifoAcquire>(
+            rewriter.getUnknownLoc(),
+            llvm::cast<LogicalObjectFifoType>(input.getType()),
+            logicalObjectFifoToConnection[input].getResult(), port);
+        auto newAccessOp = rewriter.create<AMDAIE::LogicalObjectFifoAccessOp>(
+            rewriter.getUnknownLoc(), acquireOp.getResult(), accessType);
+        rewriter.replaceAllUsesWith(accessOp.getResult(),
+                                    newAccessOp.getResult());
+
+        // Insert the release op. The location of the release is as close to the
+        // following access op as possible, but always in the same block as the
+        // access op being released.
+        AMDAIE::LogicalObjectFifoAccessOp nextAccessOp;
+        if (i + 1 != accessOps.size()) nextAccessOp = accessOps[i + 1];
+        Operation *nextAccessOpsAncestor =
+            getAncestorInBlock(nextAccessOp, block);
+        if (nextAccessOpsAncestor &&
+            nextAccessOpsAncestor->getBlock() == block) {
+          llvm::errs() << "Setting insertion point to nextAccessOpsAncestor\n";
```

Comment:
This should be removed, no?

Reply:
Oof, I clearly should have gone over this a bit more carefully after my last change. Thanks for spotting